linkerd/metrics/src/counter.rs

```diff
@@ -15,75 +15,72 @@ use super::prom::{FmtLabels, FmtMetric};
 /// [`rate()`]: https://prometheus.io/docs/prometheus/latest/querying/functions/#rate()
 /// [`irate()`]: https://prometheus.io/docs/prometheus/latest/querying/functions/#irate()
 /// [`resets()`]: https://prometheus.io/docs/prometheus/latest/querying/functions/#resets
-#[derive(Copy, Clone, Debug, Default, Eq, PartialEq)]
-pub struct Counter(u64);
-
-/// Largest `u64` that can fit without loss of precision in `f64` (2^53).
-pub(crate) const MAX_PRECISE_COUNTER: u64 = 0x20_0000_0000_0000;
+#[derive(Debug, Default)]
+pub struct Counter(AtomicU64);
 
 // ===== impl Counter =====
 
 impl Counter {
-    /// Increment the counter by one.
-    ///
-    /// This function wraps on 52-bit overflows.
-    pub fn incr(&mut self) {
-        *self += 1;
+    pub fn incr(&self) {
+        self.add(1)
+    }
+
+    pub fn add(&self, n: u64) {
+        self.0.fetch_add(n, Ordering::Release);
     }
 
-    /// Return current counter value.
+    /// Return current counter value, wrapped to be safe for use with Prometheus.
     pub fn value(&self) -> u64 {
         self.0
+            .load(Ordering::Acquire)
+            .wrapping_rem(MAX_PRECISE_VALUE + 1)
     }
 }
 
 impl Into<u64> for Counter {
     fn into(self) -> u64 {
-        self.0
+        self.value()
     }
 }
 
 impl From<u64> for Counter {
     fn from(value: u64) -> Self {
-        Counter(0) + value
+        Counter(value.into())
     }
 }
 
 impl ops::Add<u64> for Counter {
     type Output = Self;
     fn add(self, rhs: u64) -> Self::Output {
-        let wrapped = self
-            .0
-            .wrapping_add(rhs)
-            .wrapping_rem(MAX_PRECISE_COUNTER + 1);
-        Counter(wrapped)
+        self.0.fetch_add(rhs, Ordering::SeqCst);
```

Comment:
should probably be `Release`, like `add` does
```suggestion
        self.0.fetch_add(rhs, Ordering::Release);
```

Reply:
or, actually...can this just use the inherent `add` method so we don't need to keep the implementation in sync?
```suggestion
        self.add(rhs)
```